src/combinator/expression.rs

```diff
@@ -0,0 +1,340 @@
+use core::marker::PhantomData;
+
+use crate::{
+    combinator::{opt, trace},
+    error::{ErrMode, ParserError},
+    stream::{Stream, StreamIsPartial},
+    PResult, Parser,
+};
+
+use super::{empty, fail};
+
+/// Parses an expression based on operator precedence.
+#[doc(alias = "pratt")]
+#[doc(alias = "separated")]
+#[doc(alias = "shunting_yard")]
+#[doc(alias = "precedence_climbing")]
+#[inline(always)]
+pub fn expression<I, ParseOperand, O, E>(
+    start_power: i64,
+    parse_operand: ParseOperand,
+) -> Expression<
+    I,
+    O,
+    ParseOperand,
+    impl Parser<I, Prefix<I, O, E>, E>,
+    impl Parser<I, Postfix<I, O, E>, E>,
+    impl Parser<I, Infix<I, O, E>, E>,
+    E,
+>
+where
+    I: Stream + StreamIsPartial,
+    ParseOperand: Parser<I, O, E>,
+    E: ParserError<I>,
+{
+    Expression {
+        start_power,
+        parse_operand,
+        parse_prefix: fail,
+        parse_postfix: fail,
+        parse_infix: fail,
+        i: Default::default(),
+        o: Default::default(),
+        e: Default::default(),
+    }
+}
+
+pub struct Expression<I, O, ParseOperand, Pre, Post, Pix, E>
+where
+    I: Stream + StreamIsPartial,
+    ParseOperand: Parser<I, O, E>,
+    E: ParserError<I>,
+{
+    start_power: i64,
+    parse_operand: ParseOperand,
+    parse_prefix: Pre,
+    parse_postfix: Post,
+    parse_infix: Pix,
+    i: PhantomData<I>,
+    o: PhantomData<O>,
+    e: PhantomData<E>,
+}
+
+impl<I, O, ParseOperand, Pre, Post, Pix, E> Expression<I, O, ParseOperand, Pre, Post, Pix, E>
+where
+    ParseOperand: Parser<I, O, E>,
+    I: Stream + StreamIsPartial,
+    E: ParserError<I>,
+{
+    #[inline(always)]
+    pub fn prefix<NewParsePrefix>(
+        self,
+        parser: NewParsePrefix,
+    ) -> Expression<I, O, ParseOperand, NewParsePrefix, Post, Pix, E>
+    where
+        NewParsePrefix: Parser<I, Prefix<I, O, E>, E>,
+    {
+        Expression {
+            start_power: self.start_power,
+            parse_operand: self.parse_operand,
+            parse_prefix: parser,
+            parse_postfix: self.parse_postfix,
+            parse_infix: self.parse_infix,
+            i: Default::default(),
+            o: Default::default(),
+            e: Default::default(),
+        }
+    }
```

Comment:
For these, how come we don't do
```suggestion
    #[inline(always)]
    pub fn prefix<NewParsePrefix>(
        mut self,
        parser: NewParsePrefix,
    ) -> Expression<I, O, ParseOperand, NewParsePrefix, Post, Pix, E>
    where
        NewParsePrefix: Parser<I, Prefix<I, O, E>, E>,
    {
        self.parse_prefix = parser;
        self
    }
```

Reply:
Different types of the `Expression` due to the different type of the new parser